nomad/stream/ndjson.go

```diff
@@ -0,0 +1,105 @@
+package stream
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+var (
+	// NDJsonHeartbeat is the NDJson to send as a heartbeat
+	// Avoids creating many heartbeat instances
+	NDJsonHeartbeat = &structs.NDJson{Data: []byte("{}\n")}
+)
+
+// NDJsonStream is used to send new line delimited JSON and heartbeats
+// to a destination (out channel)
+type NDJsonStream struct {
+	// mu locks during send and protects running
+	mu sync.Mutex
+
+	out chan<- *structs.NDJson
+
+	// heartbeat is the interval to send heartbeat messages to keep a connection
+	// open.
+	heartbeat *time.Ticker
+
+	publishCh chan structs.NDJson
+	exitCh    chan struct{}
+
+	running bool
+}
+
+// NewNNewNDJsonStream creates a new NDJson stream that will output NDJson structs
+// to the passed output channel
+func NewNDJsonStream(out chan<- *structs.NDJson, heartbeat time.Duration) *NDJsonStream {
+	return &NDJsonStream{
+		out:       out,
+		heartbeat: time.NewTicker(heartbeat),
+		exitCh:    make(chan struct{}),
+		publishCh: make(chan structs.NDJson),
+	}
+}
+
+// Run starts a long lived goroutine that handles sending
+// heartbeats and processed json objects to the streams out channel as well
+func (n *NDJsonStream) Run(ctx context.Context) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	if n.running {
+		return
+	}
+	n.running = true
+
+	go n.run(ctx)
+}
+
+func (n *NDJsonStream) run(ctx context.Context) {
```

Comment:
What do you think about having `run` be `heartbeat` - Send can publish to out channel directly eliminating some unnecessary goroutine switch and the tracking of running.